### ISSUE TITLE
improve waiting behavior for de-provisioning external remote execution node

### ISSUE THREAD

**Issue by TheRealHaoLiu** [ansible/awx#12946]
currently in our implementation, de-provisioning a node starts a task that will wait indefinitely for all of the jobs to drain from the node 

we dont fully understand the negative path impact of this (if the job that's running is stuck)

this may potentially trigger some "timeout" causing the de-provisioning task to be reaped.

we might also want to introduce a mechanism for FORCE de-provisioning that bypass the checks

**Comment by TheRealHaoLiu** [comment]
@AlanCoding what do you think about the current behavior

**Comment by TheRealHaoLiu** [comment]
@AlanCoding agree that this can be address this issue post feature branch merge

**Comment by AlanCoding** [comment]
Yep, feature branch in, :+1: definitely want to take our time and address this properly.

> we might also want to introduce a mechanism for FORCE de-provisioning that bypass the checks

I need to get some background information here - when you tell the system to de-provision a node, does it cancel the jobs on the instance? If not, I understand, but we need to consider the definition of "force" here.

Absent other information, I assume that the node we are de-provisioning can be communicated with. If we have communication, then de-provisioning is focusing on the symptom, not the sickness. The problem is that jobs are still running, so the first-level intervention is cancel the jobs. With recent changes, canceling sends a signal to the controller node's dispatcher, which sends SIGTERM to the control process, which issues the `receptorctl` command to cancel the work unit, which becomes a SIGTERM on the runner worker process.

If anything in this process is stalled, we need to debug that particular part. Maybe lack of response to SIGTERM needs to be followed by a SIGKILL at some point. We can do things like that. If you totally lose connection to an execution node, then at some point you may need to kick it out of the mesh, which would be de-provisioning, this should just be the absolute last resort.

**Comment by AlanCoding** [comment]
Looking at this again

https://github.com/ansible/awx/blob/9c2185c68f25646f80fea5ce602540e7eb3f8e67/awx/main/tasks/receptor.py#L690

Confirming, this seems to have the intention to let the jobs follow their natural course until completion. So I would start with the idea of canceling them, and applying some appropriate follow-on wait steps.

**Comment by TheRealHaoLiu** [comment]
so currently there are these cases we need to discuss 
- waiting for the job to finish with intend of removing the node when no job is running (deprovisioning)
- canceling all job with intend of removing the node when all job are canceled (this is something that customer can manually do already)
- just delete i dont care about the currently running job just delete (forced removal)

**Comment by TheRealHaoLiu** [comment]
from @jbradberry 
### deprovisioning
- remove the wait-loop that polls for job completion
- trigger the task from within the task manager's reaper logic
- make sure that the side-effect model saves happen in an appropriate place (I think the link states will have to move out of the task)
- make sure that the task manager accumulates the nodes in 'deprovisioning' state as well as 'ready' so that the reaper can do things with them, but still don't allow jobs to run on deprovisioning nodes

**Comment by jbradberry** [comment]
PR: https://github.com/ansible/awx/pull/12980
